### DISCUSSION TITLE
What's purpose of SAFEARRAYTests

### DISCUSSION THREAD

**Discussion by kant2002** [dotnet/winforms#6756] [Q&A]
Maybe I miss something obvious, but SafeArrayTests are not testing anything.
I get that maybe SafeArray handling should be in VARIANTTests but they not there too. Do I miss something?

https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/Oleaut32/SAFEARRAYTests.cs

**Comment by RussKie** [comment]
I am not sure I can agree that these tests don't test anything. They test our definitions and implementations of [`SAFEARRAY`](https://github.com/dotnet/winforms/blob/main/src/System.Windows.Forms.Primitives/src/Interop/OleAut32/Interop.SAFEARRAY.cs) are correct.
What is your concern?

**Reply by kant2002** [reply]
For me, this tests is testing internal details of SAFEaRRAY struct, and this is all looks like tests are testing how we apply marshaling attributes.

also this line https://github.com/dotnet/winforms/blob/9f774856cf14f58be9f84d1fcfcc4eeae909140d/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/Oleaut32/SAFEARRAYTests.cs#L93

when I decide to replace built-in Com usage I have to change that line which is meh, since i dislike changing tests.

for me seems to be tests better test some other areas which are used SAFEARRAY under the hood. 

but I may miss other important context here.

**Reply by RussKie** [reply]
> also this line
> 
> https://github.com/dotnet/winforms/blob/9f774856cf14f58be9f84d1fcfcc4eeae909140d/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/Oleaut32/SAFEARRAYTests.cs#L93
> 
> when I decide to replace built-in Com usage I have to change that line which is meh, since i dislike changing tests.

Why do you think this line needs to be changed? `CustomRecordInfo` is a test private implementation of `IRecordInfo`, which isn't changing with the migration to COM Wrappers.
What am I missing?

**Reply by kant2002** [reply]
All modifications of SAFEARRAY struct happens via test only declarations

https://github.com/dotnet/winforms/blob/9f774856cf14f58be9f84d1fcfcc4eeae909140d/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/Oleaut32/SAFEARRAYTests.cs#L410-L423

Whole struct is almost regular struct with marshaling attributes applied. It is does not have behaviour. All modifications of that structs is via Win32 API. Obviously marshaling and Win32 is working. I cannot find WinForms specific functionality what is tested here. Maybe there was some historical reasons why this test appears?

SAFEARRAY struct used only to supplement Variant implementation. And these tests seems to cover more cases and do not rely on internal details too much. For example

https://github.com/dotnet/winforms/blob/9f774856cf14f58be9f84d1fcfcc4eeae909140d/src/System.Windows.Forms.Primitives/tests/UnitTests/Interop/Oleaut32/VARIANTTests.cs#L4317
